### DISCUSSION TITLE
Auto Detect Flash

### DISCUSSION THREAD

**Discussion by Icesythe7** [sanni/cartreader#469] [Coding Corner]
I've been working on adding auto detection of flash carts to the GB(C), the issue is I can't store many sequences due to available memory, this can be solved by storing everything on the sd card the question is will it be accepted like this? I don't want to waste my time if it will just be a rejected submission. 

P.S. also enabled flash carts that work on the audio pin instead of wr pin like the mbc30(not mbc3) DIY flashcarts

![image](https://user-images.githubusercontent.com/13612149/184604559-d2f072b2-0ccd-4914-a49c-1026be3b9a31.png)

```c++
#define MAX_SEQUENCES 2

constexpr word flash_sequence[MAX_SEQUENCES][7] = {
	{3, 0x5555, 0xAA, 0x02AA, 0x55, 0x0555, 0x90},
    {3, 0x0AAA, 0xA9, 0x0555, 0x56, 0x0AAA, 0x90},
};

void runSequence_GB(const uint8_t sequence_num, const bool aud)
{
    uint8_t pos = 1;
    for (uint8_t i = 0; i < flash_sequence[sequence_num][0]; ++i)
    {
        writeByte_GB(flash_sequence[sequence_num][pos], flash_sequence[sequence_num][pos + 1], aud);
        pos += 2;
    }
}

unsigned long getFlashid_GB(const bool aud)
{
    uint8_t flash_id[4];
    uint8_t rom_data[4];

    for (uint8_t i = 0; i < 4; ++i)
        rom_data[i] = readByte_GB(i);

	for (uint8_t i = 0; i < MAX_SEQUENCES; ++i)
	{
        // Run the Sequence
        runSequence_GB(i, aud);

        // Store the Sequence for comparison
        for (uint8_t x = 0; x < 4; ++x)
            flash_id[x] = readByte_GB(x);

        if (!isEqual(rom_data, flash_id))
            return (static_cast<unsigned long>(flash_id[0]) << 24 | static_cast<unsigned long>(flash_id[1]) << 16 |
                static_cast<unsigned long>(flash_id[2]) << 8 | static_cast<unsigned long>(flash_id[3]));
	}

    return 0;
}`

**Comment by sanni** [comment]
Sure. 
If it's just an additional way of flashing no worries. If it replaces the current way you can #define an option and use #ifdef so that the old way of flashing is still accessible as a fall back in case it is needed some time.

**Reply by Icesythe7** [reply]
The old way of manually selecting your flash chip would be obsolete yes as it will detect it automatically, on top of that the entire gb.ino needs a rewrite that I am working on as alot of the timings are off/incorrect and alot of program space can be reclaimed aswell by moving static data to the SD, the timing issue causes instability issues in some carts especially while flashing not yet implemented carts.

**Reply by sanni** [reply]
Sounds good to me. The 4 different flash options GB.ino has right now is confusing anyway. So it could use some overhauling. 👍

**Reply by Icesythe7** [reply]
Yes because the AmF016 you have options for mbc3 and mbc5 but the mbc doesn't matter so it's a waste of coding space anyways since both are the exact same.
